File: backups/20251126_214534/scripts/generate_sitemap.py
#!/usr/bin/env python3
"""
Sitemap Generator for Ideas Before Time
Generates sitemap.xml from innovations.json
"""

import json
import os
from datetime import datetime
from pathlib import Path

def load_innovations():
    """Load innovations from JSON file"""
    json_path = Path(__file__).parent.parent / 'innovations.json'
    with open(json_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def generate_sitemap():
    """Generate sitemap.xml from innovations data"""
    data = load_innovations()
    innovations = data.get('innovations', [])
    
    # Start sitemap
    sitemap = ['<?xml version="1.0" encoding="UTF-8"?>']
    sitemap.append('<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">')
    
    # Base URL
    base_url = 'https://ideas-before-time.pages.dev'
    today = datetime.now().strftime('%Y-%m-%d')
    
    # Static pages
    static_pages = [
        ('/', '1.0', 'daily'),
        ('/library.html', '0.9', 'weekly'),
        ('/about.html', '0.8', 'monthly'),
        ('/timeline.html', '0.8', 'monthly'),
        ('/innovations-table.html', '0.9', 'weekly'),
        ('/business-insights-enhanced.html', '0.7', 'monthly'),
        ('/specialty-chemicals-index.html', '0.7', 'monthly'),
    ]
    
    for url, priority, changefreq in static_pages:
        sitemap.append('  <url>')
        sitemap.append(f'    <loc>{base_url}{url}</loc>')
        sitemap.append(f'    <lastmod>{today}</lastmod>')
        sitemap.append(f'    <changefreq>{changefreq}</changefreq>')
        sitemap.append(f'    <priority>{priority}</priority>')
        sitemap.append('  </url>')
    
    # Innovation pages
    for innovation in innovations:
        if innovation.get('status') == 'published':
            innovation_id = innovation.get('id')
            year = innovation.get('year', 2000)
            
            # Calculate priority based on year (newer = higher)
            current_year = datetime.now().year
            years_old = current_year - year
            priority = max(0.5, min(0.9, 0.9 - (years_old * 0.02)))
            
            sitemap.append('  <url>')
            sitemap.append(f'    <loc>{base_url}/{innovation_id}.html</loc>')
            sitemap.append(f'    <lastmod>{today}</lastmod>')
            sitemap.append(f'    <changefreq>monthly</changefreq>')
            sitemap.append(f'    <priority>{priority:.1f}</priority>')
            sitemap.append('  </url>')
    
    sitemap.append('</urlset>')
    
    # Write sitemap
    sitemap_path = Path(__file__).parent.parent / 'sitemap.xml'
    with open(sitemap_path, 'w', encoding='utf-8') as f:
        f.write('\n'.join(sitemap))
    
    print(f"✅ Sitemap generated successfully!")
    print(f"   Total URLs: {len(static_pages) + len([i for i in innovations if i.get('status') == 'published'])}")
    print(f"   Output: {sitemap_path}")

if __name__ == '__main__':
    try:
        generate_sitemap()
    except Exception as e:
        print(f"❌ Error generating sitemap: {e}")
        exit(1)

File: backups/20251126_214534/scripts/generate-sitemap.js
/**
 * Sitemap Generator for Ideas Before Time
 * Reads innovations.json and generates sitemap.xml automatically
 */

const fs = require('fs');
const path = require('path');

// Configuration
const BASE_URL = 'https://ideas-before-time.pages.dev';
const INNOVATIONS_FILE = path.join(__dirname, '../innovations.json');
const SITEMAP_FILE = path.join(__dirname, '../sitemap.xml');

// Static pages with their priorities
const STATIC_PAGES = [
    { url: '/', priority: 1.0, changefreq: 'monthly' },
    { url: '/library.html', priority: 0.9, changefreq: 'weekly' },
    { url: '/about.html', priority: 0.9, changefreq: 'monthly' },
    { url: '/timeline.html', priority: 0.8, changefreq: 'monthly' },
    { url: '/cv-preview.html', priority: 0.8, changefreq: 'monthly' },
    { url: '/business-insights.html', priority: 0.85, changefreq: 'weekly' },
    { url: '/teaching-ai-to-think.html', priority: 0.9, changefreq: 'weekly' },
    { url: '/specialty-chemicals-index.html', priority: 0.8, changefreq: 'monthly' },
    { url: '/specialty-chemicals-qa.html', priority: 0.8, changefreq: 'monthly' },
    { url: '/specialty-chemicals-value-chain.html', priority: 0.8, changefreq: 'monthly' }
];

/**
 * Read and parse innovations.json
 */
function readInnovations() {
    try {
        const data = fs.readFileSync(INNOVATIONS_FILE, 'utf8');
        return JSON.parse(data);
    } catch (error) {
        console.error('Error reading innovations.json:', error.message);
        process.exit(1);
    }
}

/**
 * Generate sitemap XML content
 */
function generateSitemap(innovations) {
    const today = new Date().toISOString().split('T')[0];
    
    let xml = '<?xml version="1.0" encoding="UTF-8"?>\n';
    xml += '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n';
    
    // Add static pages
    xml += '  <!-- Main Pages -->\n';
    STATIC_PAGES.forEach(page => {
        xml += '  <url>\n';
        xml += `    <loc>${BASE_URL}${page.url}</loc>\n`;
        xml += `    <lastmod>${today}</lastmod>\n`;
        xml += `    <priority>${page.priority}</priority>\n`;
        xml += `    <changefreq>${page.changefreq}</changefreq>\n`;
        xml += '  </url>\n';
    });
    
    // Add innovation pages
    xml += '  \n  <!-- Innovation Pages -->\n';
    innovations.innovations
        .filter(innovation => innovation.status === 'published')
        .forEach(innovation => {
            xml += '  <url>\n';
            xml += `    <loc>${BASE_URL}/${innovation.id}.html</loc>\n`;
            xml += `    <lastmod>${today}</lastmod>\n`;
            xml += `    <priority>0.7</priority>\n`;
            xml += `    <changefreq>yearly</changefreq>\n`;
            xml += '  </url>\n';
        });
    
    xml += '</urlset>\n';
    
    return xml;
}

/**
 * Write sitemap to file
 */
function writeSitemap(xml) {
    try {
        fs.writeFileSync(SITEMAP_FILE, xml, 'utf8');
        console.log('✓ Sitemap generated successfully');
        console.log(`  Location: ${SITEMAP_FILE}`);
    } catch (error) {
        console.error('Error writing sitemap:', error.message);
        process.exit(1);
    }
}

/**
 * Main execution
 */
function main() {
    console.log('Generating sitemap...');
    
    const innovations = readInnovations();
    console.log(`  Found ${innovations.innovations.length} innovations`);
    
    const xml = generateSitemap(innovations);
    console.log(`  Generated ${xml.split('<url>').length - 1} URLs`);
    
    writeSitemap(xml);
}

// Run if called directly
if (require.main === module) {
    main();
}

module.exports = { generateSitemap, readInnovations };
